### PR TITLE
Replace underscore by lodash

### PR DESCRIPTION
**Pull request by Kureev**
As far as we agreed to merge `rnpm` into react-native core, we need to align our dependencies to exclude duplications. One of the steps forward would be to use the same utilities library. According to the thread on fb, everybody is fine with replacing underscore by lodash (which we use internally for rnpm). 

So, here we go!

cc @mkonicek @davidaurelio @grabbou 

**Test plan**
```
$ npm test
```
image

**Code formatting**

Changes are aligned with the current [style guide](https://github.com/facebook/react-native/blob/master/CONTRIBUTING.md#style-guide).

### CODE REVIEW

**Comment by grabbou** [comment]
Thanks, looks good to me. I am pretty sure you will have to squash commits, but you can probably wait till someone confirms nothing to be changed here.

**Comment by Kureev** [comment]
@grabbou Yeah, I keep them separated for now (easier to revert if needed), but later on I can squash them into one.

**Comment by satya164** [comment]
@grabbou 

> I am pretty sure you will have to squash commits

The bot will squash commits when this gets merged :)

**Comment by bestander** [comment]
@satya164, when it is ready, instead of `shipit` do `import`, and ping me.
I'll need to update npm deps internally

**Comment by satya164** [comment]
@bestander Okies. I will :)

**Comment by satya164** [comment]
Can we require individual modules instead of the whole package? For example `require('lodash/flattenDeep')`?

Doesn't make any difference since it's just the packager though.

**Comment by gaearon** [comment]
>Can we require individual modules instead of the whole package? For example require('lodash/flattenDeep')?

Agree that this would be nicer. Less code to process is better :+1:

File: packager/react-packager/index.js [hunk 1]
```diff
@@ -118,6 +117,16 @@ function createNonPersistentServer(options) {
   return createServer(options);
 }
 
+function omit(obj, blacklistedKeys) {
```
Comment (satya164) [inline packager/react-packager/index.js]: Why not use `lodash/omit` ?

**Comment by satya164** [comment]
Looks good overall. This is awesome, because lots of other packages are using `lodash` anyways. The less deps, the better.

**Comment by Kureev** [comment]
I agree about individual modules, but I remember in the facebook thread @davidaurelio had some objections about it. 

@satya164 Ideally, I'd like to replace `lodash` by self-written functions (to eliminate one of the ext. dependencies). But I think it shouldn't be approached in this PR and potentially requires more communication.

File: packager/react-packager/index.js (hunk 1 above)
Comment (Kureev) [inline packager/react-packager/index.js]: Ideally, I'd like to replace `lodash` by a self-written functions (to eliminate one of the ext. dependencies). But I think it shouldn't approach it in this PR.

Comment (grabbou) [inline packager/react-packager/index.js]: I think i'd be nice to just add a small test suite for that then.

**Comment by satya164** [comment]
@Kureev 

> I agree about individual modules, but I remember in the facebook thread @davidaurelio had some objections about it.

Let's wait for @davidaurelio 's feedback then.

> @satya164 Ideally, I'd like to replace lodash by self-written functions (to eliminate one of the ext. dependencies).

IMO it's not worth it, `lodash` is used by lots of packages. So we're not really getting rid of it.

File: packager/react-packager/index.js (hunk 1 above)
Comment (Kureev) [inline packager/react-packager/index.js]: @grabbou that would mean that we need to extract this function somewhere. I'd like to do it, but I have no clue where to put it. Any suggestions?

**Comment by Kureev** [comment]
@satya164 maybe you're right. But as you said, let's wait for the @davidaurelio's feedback first :+1:

**Comment by mkonicek** [comment]
Thank you for doing this!

**Comment by satya164** [comment]
ping @davidaurelio

**Comment by mkonicek** [comment]
David is very busy trying to merge the PRs that will unbreak Travis. Let's please hold on with this until Travis is green. Thanks for the patience!

**Comment by satya164** [comment]
ping @davidaurelio

**Comment by Kureev** [comment]
I've rebased this branch by master. Now tests should be green :green_apple:

**Comment by davidaurelio** [comment]
@facebook-github-bot shipit

**Comment by davidaurelio** [comment]
I will hold this until Monday. It will get in the next version.

**Comment by Kureev** [comment]
:+1:

**Comment by mkonicek** [comment]
Hey guys, I'm merging https://github.com/facebook/react-native/pull/6346 and then this one right after.

**Comment by davidaurelio** [comment]
thanks @mkonicek

**Comment by mkonicek** [comment]
OK #6346 is landing internally now, once it lands want to rebase this one on top of it because both require updating the `npm-shrinkwrap.json` file. It's late in London now so will continue in the morning. Thanks for the patience!

**Comment by Kureev** [comment]
ping me when I can rebase

**Comment by mkonicek** [comment]
Turns out https://github.com/facebook/react-native/pull/6346 doesn't work with our internal build based on Buck so we can probably merge this one first.

**Comment by mkonicek** [comment]
(oops wrong button :smiley: )

**Comment by mkonicek** [comment]
OK I did some investigation and this PR is in fact blocked on https://github.com/facebook/react-native/pull/6346. When I try merging this PR on its own and regenerate `npm-shrinkwrap.json` the packager doesn't work when used with npm 2.

The root of the problem is we can't regenerate `npm-shrinkwrap.json` in a way that works with npm 2, and since changes to `package.json` require `npm-shrinkwrap.json` to be regenerated, we can't merge any PRs that change `package.json` until https://github.com/facebook/react-native/pull/6346 lands.

**Comment by Kureev** [comment]
image

**Comment by davidaurelio** [comment]
Can we get away with lodash 3? babel depends on that. In that case, `npm install` puts three versions of lodash in the tree. If we need v4, we end up with 20.

**Comment by davidaurelio** [comment]
no need to update the PR. I have it updated internally

**Comment by Kureev** [comment]
@davidaurelio sure, thanks! I think it should work even without any code changes (except the lodash version in package.json). 

@grabbou we probably need to follow up on this in the link and install plugins (just to be sure that it works with `lodash@3` without any code changes).

**Comment by grabbou** [comment]
Yeah, let's apply updates to our internal packages once this is merged.

**Pull request closed without merging**